File: lib/analytics/inventory.ts
/**
 * Складская аналитика по формулам исходной таблицы ОМБОР аналитикаси.
 * Использует исправленные диапазоны и добавляет построчный расчёт замороженных денег.
 */

export type InventoryInput = {
  id: string;
  name: string;
  stock: number;          // C: остаток, шт
  costPrice: number;      // D: закупочная цена
  salePrice: number;      // E: цена продажи
  avgDailySales: number;  // F: средние продажи в день, шт
  normDays: number;       // G: норматив запаса, дней
};

export type InventoryRow = InventoryInput & {
  stockValue: number;      // H: остаток в деньгах = stock * costPrice
  stockDays: number;       // I: запас в днях = stock / avgDailySales
  dailyGross: number;      // J: дневной валовый доход = (sale-cost)*avgDailySales
  share: number;           // K: доля в общем валовом доходе
  margin: number;          // M: маржа = (sale-cost)/sale
  markup: number;          // N: наценка = (sale-cost)/cost
  oosLoss: number;         // O: потери из-за OOS (если stockDays < normDays)
  frozenMoney: number;     // P: замороженные деньги (если stockDays > normDays)
  frozenShare: number;     // Q: доля замороженных от стоимости остатка
  mmlFlag: boolean;        // L: попал ли товар в MML
};

export type InventoryTotals = {
  stockValue: number;
  dailyGross: number;
  mmlShare: number;        // суммарная доля MML товаров
  oosLoss: number;
  frozenMoney: number;
  frozenShare: number;
  // Бизнес-итог на горизонте N дней
  horizonDays: number;
  potentialProfit: number;  // максимально возможная прибыль за horizon
  lostProfit: number;       // потери из-за OOS за horizon
  actualProfit: number;     // фактическая прибыль за horizon
  profitUpliftPct: number;  // на сколько % выросла бы прибыль при OOS=0
  profitUpliftX: number;    // во сколько раз выросла бы прибыль
};

export type InventoryReport = {
  rows: InventoryRow[];
  totals: InventoryTotals;
};

export type InventoryOptions = {
  horizonDays?: number;     // горизонт расчёта (по умолчанию 10 как в исходной таблице)
  /**
   * Кумулятивная доля выручки для попадания в MML.
   * Сортируем товары по share по убыванию и помечаем верхние, пока их
   * суммарная доля не достигнет порога. По умолчанию 0.80 (≈ABC класс A).
   * Раньше использовалось per-item threshold 15% — это работало только
   * на маленьких каталогах (10-50 SKU). На каталоге в 1000+ SKU ни один
   * товар по отдельности не наберёт 15%, и MML всегда был пустым.
   */
  mmlCumulativeShare?: number;
};

export function buildInventoryReport(
  inputs: InventoryInput[],
  opts: InventoryOptions = {},
): InventoryReport {
  const horizonDays = opts.horizonDays ?? 10;
  const mmlCum = opts.mmlCumulativeShare ?? 0.80;

  const partial = inputs.map((r) => {
    const stockValue = r.stock * r.costPrice;
    const stockDays = r.avgDailySales > 0 ? r.stock / r.avgDailySales : Infinity;
    const dailyGross = (r.salePrice - r.costPrice) * r.avgDailySales;
    const margin = r.salePrice > 0 ? (r.salePrice - r.costPrice) / r.salePrice : 0;
    const markup = r.costPrice > 0 ? (r.salePrice - r.costPrice) / r.costPrice : 0;

    // OOS только если запас МЕНЬШЕ норматива (avgDailySales=0 → stockDays=∞ → false → 0)
    const oosLoss =
      stockDays < r.normDays ? (r.normDays - stockDays) * dailyGross : 0;
    // Замороженные деньги = всё, что лежит на складе сверх норматива.
    // Эквивалентная формула: stockValue - normDays * avgDailySales * costPrice
    // (избегает Infinity * 0 = NaN, когда avgDailySales = 0)
    const frozenMoney = Math.max(
      0,
      stockValue - r.normDays * r.avgDailySales * r.costPrice,
    );
    const frozenShare = stockValue > 0 ? frozenMoney / stockValue : 0;

    return {
      ...r,
      stockValue,
      stockDays,
      dailyGross,
      margin,
      markup,
      oosLoss,
      frozenMoney,
      frozenShare,
    };
  });

  const totalDailyGross = partial.reduce((s, x) => s + x.dailyGross, 0);

  const withShare = partial.map((r) => ({
    ...r,
    share: totalDailyGross > 0 ? r.dailyGross / totalDailyGross : 0,
  }));

  // Помечаем верхние товары по выручке, пока их суммарная доля не достигнет порога
  const mmlIds = new Set<string>();
  if (mmlCum > 0 && totalDailyGross > 0) {
    const sorted = [...withShare].sort((a, b) => b.share - a.share);
    let cum = 0;
    for (const r of sorted) {
      if (cum >= mmlCum) break;
      if (r.share <= 0) break; // не помечаем товары без продаж
      mmlIds.add(r.id);
      cum += r.share;
    }
  }

  const rows: InventoryRow[] = withShare.map((r) => ({
    ...r,
    mmlFlag: mmlIds.has(r.id),
  }));

  const stockValueT = rows.reduce((s, x) => s + x.stockValue, 0);
  const oosLossT = rows.reduce((s, x) => s + x.oosLoss, 0);
  const frozenT = rows.reduce((s, x) => s + x.frozenMoney, 0);
  const mmlShare = rows.filter((r) => r.mmlFlag).reduce((s, x) => s + x.share, 0);

  const potentialProfit = totalDailyGross * horizonDays;
  const lostProfit = oosLossT;
  const actualProfit = potentialProfit - lostProfit;
  const profitUpliftPct = actualProfit > 0 ? lostProfit / actualProfit : 0;
  const profitUpliftX = actualProfit > 0 ? potentialProfit / actualProfit : 0;

  return {
    rows,
    totals: {
      stockValue: stockValueT,
      dailyGross: totalDailyGross,
      mmlShare,
      oosLoss: oosLossT,
      frozenMoney: frozenT,
      frozenShare: stockValueT > 0 ? frozenT / stockValueT : 0,
      horizonDays,
      potentialProfit,
      lostProfit,
      actualProfit,
      profitUpliftPct,
      profitUpliftX,
    },
  };
}

File: backend/app/analytics/inventory.py
"""
Складская аналитика — порт widget/lib/analytics/inventory.ts.

Формулы один-в-один с Excel-таблицей ОМБОР аналитикаси, на которые
опирается клиент:
  H stockValue   = stock × costPrice
  I stockDays    = stock / avgDailySales   (∞ если avgDailySales == 0)
  J dailyGross   = (salePrice − costPrice) × avgDailySales
  K share        = dailyGross / SUM(dailyGross)
  L mmlFlag      = TRUE для топ-товаров пока их кумулятивная доля ≤ mml_cumulative_share
  M margin       = (salePrice − costPrice) / salePrice
  N markup       = (salePrice − costPrice) / costPrice
  O oosLoss      = (normDays − stockDays) × dailyGross   (только если stockDays < normDays)
  P frozenMoney  = max(0, stockValue − normDays × avgDailySales × costPrice)
  Q frozenShare  = frozenMoney / stockValue

Бизнес-итог на горизонте N дней:
  potentialProfit = SUM(dailyGross) × horizon
  lostProfit      = SUM(oosLoss)
  actualProfit    = potentialProfit − lostProfit
  profitUpliftPct = lostProfit / actualProfit      (если actualProfit > 0)
  profitUpliftX   = potentialProfit / actualProfit (если actualProfit > 0)
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class InventoryInput:
    id: str
    name: str
    stock: float
    cost_price: float
    sale_price: float
    avg_daily_sales: float
    norm_days: float
    buy_currency: Optional[str] = None
    sale_currency: Optional[str] = None
    cost_price_original: Optional[float] = None
    sale_price_original: Optional[float] = None
    converted: bool = False
    cost_from_fifo: bool = False


@dataclass
class InventoryRow:
    id: str
    name: str
    stock: float
    cost_price: float
    sale_price: float
    avg_daily_sales: float
    norm_days: float
    stock_value: float
    stock_days: float
    daily_gross: float
    share: float
    margin: float
    markup: float
    oos_loss: float
    frozen_money: float
    frozen_share: float
    mml_flag: bool
    buy_currency: Optional[str] = None
    sale_currency: Optional[str] = None
    cost_price_original: Optional[float] = None
    sale_price_original: Optional[float] = None
    converted: bool = False
    cost_from_fifo: bool = False


@dataclass
class InventoryTotals:
    stock_value: float
    daily_gross: float
    mml_share: float
    oos_loss: float
    frozen_money: float
    frozen_share: float
    horizon_days: float
    potential_profit: float
    lost_profit: float
    actual_profit: float
    profit_uplift_pct: float
    profit_uplift_x: float


@dataclass
class InventoryReport:
    rows: list[InventoryRow]
    totals: InventoryTotals


def build_inventory_report(
    inputs: list[InventoryInput],
    horizon_days: float = 10,
    mml_cumulative_share: float = 0.80,
) -> InventoryReport:
    partial: list[dict] = []
    for r in inputs:
        stock_value = r.stock * r.cost_price
        stock_days = (r.stock / r.avg_daily_sales) if r.avg_daily_sales > 0 else math.inf
        daily_gross = (r.sale_price - r.cost_price) * r.avg_daily_sales
        margin = ((r.sale_price - r.cost_price) / r.sale_price) if r.sale_price > 0 else 0.0
        markup = ((r.sale_price - r.cost_price) / r.cost_price) if r.cost_price > 0 else 0.0

        oos_loss = ((r.norm_days - stock_days) * daily_gross) if stock_days < r.norm_days else 0.0
        # Эквивалентная формула — избегает Infinity * 0 = NaN когда avg_daily_sales = 0
        frozen_money = max(0.0, stock_value - r.norm_days * r.avg_daily_sales * r.cost_price)
        frozen_share = (frozen_money / stock_value) if stock_value > 0 else 0.0

        partial.append({
            "input": r,
            "stock_value": stock_value,
            "stock_days": stock_days,
            "daily_gross": daily_gross,
            "margin": margin,
            "markup": markup,
            "oos_loss": oos_loss,
            "frozen_money": frozen_money,
            "frozen_share": frozen_share,
        })

    total_daily_gross = sum(p["daily_gross"] for p in partial)

    for p in partial:
        p["share"] = (p["daily_gross"] / total_daily_gross) if total_daily_gross > 0 else 0.0

    # MML: топ товары по выручке, пока кумулятивная доля не достигнет порога
    mml_ids: set[str] = set()
    if mml_cumulative_share > 0 and total_daily_gross > 0:
        sorted_by_share = sorted(partial, key=lambda x: x["share"], reverse=True)
        cum = 0.0
        for p in sorted_by_share:
            if cum >= mml_cumulative_share:
                break
            if p["share"] <= 0:
                break
            mml_ids.add(p["input"].id)
            cum += p["share"]

    rows: list[InventoryRow] = []
    for p in partial:
        r = p["input"]
        rows.append(InventoryRow(
            id=r.id, name=r.name,
            stock=r.stock, cost_price=r.cost_price, sale_price=r.sale_price,
            avg_daily_sales=r.avg_daily_sales, norm_days=r.norm_days,
            stock_value=p["stock_value"],
            stock_days=p["stock_days"],
            daily_gross=p["daily_gross"],
            share=p["share"],
            margin=p["margin"],
            markup=p["markup"],
            oos_loss=p["oos_loss"],
            frozen_money=p["frozen_money"],
            frozen_share=p["frozen_share"],
            mml_flag=r.id in mml_ids,
            buy_currency=r.buy_currency, sale_currency=r.sale_currency,
            cost_price_original=r.cost_price_original,
            sale_price_original=r.sale_price_original,
            converted=r.converted, cost_from_fifo=r.cost_from_fifo,
        ))

    stock_value_t = sum(x.stock_value for x in rows)
    oos_loss_t = sum(x.oos_loss for x in rows)
    frozen_t = sum(x.frozen_money for x in rows)
    mml_share = sum(x.share for x in rows if x.mml_flag)

    potential_profit = total_daily_gross * horizon_days
    lost_profit = oos_loss_t
    actual_profit = potential_profit - lost_profit
    profit_uplift_pct = (lost_profit / actual_profit) if actual_profit > 0 else 0.0
    profit_uplift_x = (potential_profit / actual_profit) if actual_profit > 0 else 0.0

    return InventoryReport(
        rows=rows,
        totals=InventoryTotals(
            stock_value=stock_value_t,
            daily_gross=total_daily_gross,
            mml_share=mml_share,
            oos_loss=oos_loss_t,
            frozen_money=frozen_t,
            frozen_share=(frozen_t / stock_value_t) if stock_value_t > 0 else 0.0,
            horizon_days=horizon_days,
            potential_profit=potential_profit,
            lost_profit=lost_profit,
            actual_profit=actual_profit,
            profit_uplift_pct=profit_uplift_pct,
            profit_uplift_x=profit_uplift_x,
        ),
    )
